### ISSUE TITLE
More messages all over in feature benchmark

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#25383]
### What version of Materialize are you using?

765ff86ea5

### What is the issue?

Seen in https://buildkite.com/materialize/nightlies/builds/6538#018dc2a4-9fb4-4486-9c03-9b132bfc2640 in all feature benchmark runs. Example:
```
2024-02-19 20:14:24 UTC	Benchmark Report for cycle 10:
2024-02-19 20:14:24 UTC	NAME                                | TYPE      |      THIS       |      OTHER      |  Regression?  | 'THIS' is:
2024-02-19 20:14:24 UTC	----------------------------------------------------------------------------------------------------
2024-02-19 20:14:24 UTC	FastPathFilterIndex                 | wallclock |           1.779 |           1.743 |      no       | 2.1 pct   more/slower
2024-02-19 20:14:24 UTC	FastPathFilterIndex                 | messages  |    31264553.000 |    27264320.000 |    !!YES!!    | 14.7 pct   more/slower
2024-02-19 20:14:24 UTC	FastPathFilterIndex                 | memory    |         996.590 |         975.609 |      no       | 2.2 pct   more/slower
2024-02-19 20:14:24 UTC	SkewedJoin                          | wallclock |           3.735 |           3.760 |      no       | 0.7 pct   less/faster
2024-02-19 20:14:24 UTC	SkewedJoin                          | messages  |    14336685.000 |    10112792.000 |    !!YES!!    | 41.8 pct   more/slower
2024-02-19 20:14:24 UTC	SkewedJoin                          | memory    |        1167.297 |        1176.834 |      no       | 0.8 pct   less/faster
```
Many benchmarks are using more messages (the sum of all messages in the system from [mz_internal.mz](http://mz_internal.mz/)_message_counts_per_worker)

> The `mz_message_counts` view describes the messages and message batches sent and received over the [dataflow] channels in the system.

The changes in the range are https://github.com/MaterializeInc/materialize/compare/aada4946ee...765ff86ea5

The most suspicious looking is https://github.com/MaterializeInc/materialize/pull/24906 @teskje 
Slightly less suspicious: https://github.com/MaterializeInc/materialize/pull/25374 @antiguru

Is this an expected regression? Does it make sense to measure the message counts in this way?

This would be a regression for v0.89

**Comment by teskje** [comment]
Yes, it is expected: https://github.com/MaterializeInc/materialize/pull/24906#issuecomment-1927190355

> Does it make sense to measure the message counts in this way?

Message counts are likely to increase whenever we render new operators and therefore more channels. If we want to measure message counts then I think this is a good way, yes. I'm less sure about the usefulness of measuring them. Was this introduced to catch a specific kind of performance regression?

**Comment by def-** [comment]
I'm not sure, I think Philip added them.

**Comment by antiguru** [comment]
> Was this introduced to catch a specific kind of performance regression?

Yes -- I think we had two problems: outer joins without consolidates, which means the records and the negated records stayed around for longer than necessary, and this would show up in the number of messages, and operators outputting a message for each record instead of batching data.
